batch/sql/estimated-current.sql

```diff
@@ -645,6 +693,19 @@ BEGIN
   END IF;
 END $$
 
+DROP TRIGGER IF EXISTS batches_after_update $$
+CREATE TRIGGER batches_after_update AFTER UPDATE ON batches
+FOR EACH ROW
+BEGIN
+  IF OLD.migrated_batch = 0 AND NEW.migrated_batch = 1 THEN
+    INSERT INTO job_groups (batch_id, job_group_id, `user`, cancel_after_n_failures, `state`, n_jobs, time_created, time_completed, callback, attributes)
+    VALUES (NEW.id, 0, NEW.`user`, NEW.cancel_after_n_failures, NEW.state, NEW.n_jobs, NEW.time_created, NEW.time_completed, NEW.callback, NEW.attributes);
+
+    INSERT INTO job_group_parents (batch_id, job_group_id, parent_id, `level`)
+    VALUES (NEW.id, 0, 0, 0);
```

Comment:
I'm a bit surprised by this. Does this substantially simplify other code? Having a self-loop in the parents table is confusing because it violates the statement

> if `(_, y, z, _)` is in `job_groups_parents` then `z` is the parent of `y`.

Reply:
I believe if we don't do this, then we'll need to split the queries into two in a bunch of places or maybe use loops. Once to do the operation on the group of interest and then another to do it on all its parents. Here's an example of billing propagation up the tree and then cancellation down the tree:

```sql
    INSERT INTO aggregated_batch_resources_v3 (batch_id, job_group_id, resource_id, token, `usage`)
    SELECT attempt_resources.batch_id,
      job_group_parents.parent_id,
      attempt_resources.deduped_resource_id,
      rand_token,
      msec_diff_rollup * quantity
    FROM attempt_resources
    LEFT JOIN jobs ON attempt_resources.batch_id = jobs.batch_id AND attempt_resources.job_id = jobs.job_id
    LEFT JOIN job_group_parents ON jobs.batch_id = job_group_parents.batch_id AND jobs.job_group_id = job_group_parents.job_group_id
    JOIN aggregated_batch_resources_v2 ON
      aggregated_batch_resources_v2.batch_id = attempt_resources.batch_id AND
      aggregated_batch_resources_v2.resource_id = attempt_resources.resource_id AND
      aggregated_batch_resources_v2.token = rand_token
    WHERE attempt_resources.batch_id = NEW.batch_id AND attempt_resources.job_id = NEW.job_id AND attempt_resources.attempt_id = NEW.attempt_id AND migrated = 1
    ON DUPLICATE KEY UPDATE `usage` = aggregated_batch_resources_v3.`usage` + msec_diff_rollup * quantity;
```

```sql
  INSERT INTO batch_inst_coll_cancellable_resources (batch_id, update_id, job_group_id, inst_coll, token,
    n_ready_cancellable_jobs, ready_cancellable_cores_mcpu, n_creating_cancellable_jobs, n_running_cancellable_jobs,
    running_cancellable_cores_mcpu)
  SELECT batch_inst_coll_cancellable_resources.batch_id, batch_inst_coll_cancellable_resources.update_id, job_group_parents.parent_id, inst_coll, 0,
    -1 * (@n_ready_cancellable_jobs := COALESCE(SUM(n_ready_cancellable_jobs), 0)),
    -1 * (@ready_cancellable_cores_mcpu := COALESCE(SUM(ready_cancellable_cores_mcpu), 0)),
    -1 * (@n_creating_cancellable_jobs := COALESCE(SUM(n_creating_cancellable_jobs), 0)),
    -1 * (@n_running_cancellable_jobs := COALESCE(SUM(n_running_cancellable_jobs), 0)),
    -1 * (@running_cancellable_cores_mcpu := COALESCE(SUM(running_cancellable_cores_mcpu), 0))
  FROM batch_inst_coll_cancellable_resources
  INNER JOIN batch_updates ON batch_inst_coll_cancellable_resources.batch_id = batch_updates.batch_id AND
    batch_inst_coll_cancellable_resources.update_id = batch_updates.update_id
  INNER JOIN job_group_parents ON batch_inst_coll_cancellable_resources.batch_id = job_group_parents.batch_id AND
    batch_inst_coll_cancellable_resources.job_group_id = job_group_parents.job_group_id
  WHERE job_group_parents.batch_id = in_batch_id AND
    job_group_parents.job_group_id = in_job_group_id AND
    batch_updates.committed
  GROUP BY batch_inst_coll_cancellable_resources.batch_id, batch_inst_coll_cancellable_resources.update_id,
    job_group_parents.parent_id, inst_coll
  ON DUPLICATE KEY UPDATE
    n_ready_cancellable_jobs = n_ready_cancellable_jobs - @n_ready_cancellable_jobs,
    ready_cancellable_cores_mcpu = ready_cancellable_cores_mcpu - @ready_cancellable_cores_mcpu,
    n_creating_cancellable_jobs = n_creating_cancellable_jobs - @n_creating_cancellable_jobs,
    n_running_cancellable_jobs = n_running_cancellable_jobs - @n_running_cancellable_jobs,
    running_cancellable_cores_mcpu = running_cancellable_cores_mcpu - @running_cancellable_cores_mcpu;

  # delete all rows that are children of this job group
  DELETE batch_inst_coll_cancellable_resources FROM batch_inst_coll_cancellable_resources
  LEFT JOIN batch_updates ON batch_inst_coll_cancellable_resources.batch_id = batch_updates.batch_id AND
    batch_inst_coll_cancellable_resources.update_id = batch_updates.update_id
  INNER JOIN job_group_parents ON batch_inst_coll_cancellable_resources.batch_id = job_group_parents.batch_id AND
    batch_inst_coll_cancellable_resources.job_group_id = job_group_parents.job_group_id
  WHERE batch_inst_coll_cancellable_resources.batch_id = in_batch_id AND
    job_group_parents.parent_id = in_job_group_id AND
    batch_updates.committed;
```

Reply:
There might be other queries in the client code that having the self loop is helpful.

Reply:
This is I believe the major outstanding point of discussion.

Reply:
Ah, I think this table has the wrong name. Knowing only the name, I thought this was true:

> if `(_, y, z, _)` is in `job_groups_parents` then `z` is the parent of `y`.

But after seeing the self-loop I thought this was true:

> if `(_, y, z, _)` is in `job_groups_parents` then `z` is the parent of `y` or `z` == `y`.

But now I see that what this *actually* is is something like `job_groups_needing_to_be_updated_when_a_job_in_this_group_changes`. It's not the ancestors (because of the self loop) and it's not the parents (because it includes ancestors and a self loop).

We need a clear and correct name for this table, but I do not object to the table itself.

Reply:
How about `job_group_tree`?